src/adapter/src/session.rs

```diff
@@ -102,28 +102,43 @@ impl<T: CoordTimestamp> Session<T> {
         wall_time: DateTime<Utc>,
         access: Option<TransactionAccessMode>,
         isolation_level: Option<TransactionIsolationLevel>,
-    ) -> Self {
+    ) -> (Self, Result<(), AdapterError>) {
+        if let Some(txn) = self.transaction.inner() {
+            if matches!(txn.ops, TransactionOps::Peeks(_))
+                && access == Some(TransactionAccessMode::ReadWrite)
+                && txn.access == Some(TransactionAccessMode::ReadOnly)
+            {
+                return (self, Err(AdapterError::ReadWriteUnavailable));
+            }
+        }
+
         match self.transaction {
-            TransactionStatus::Default | TransactionStatus::Started(_) => {
+            TransactionStatus::Default => {
                 self.transaction = TransactionStatus::InTransaction(Transaction {
                     pcx: PlanContext::new(wall_time, self.vars.qgm_optimizations()),
                     ops: TransactionOps::None,
                     write_lock_guard: None,
                     access,
                 });
             }
-            TransactionStatus::InTransactionImplicit(txn) => {
+            TransactionStatus::Started(mut txn)
+            | TransactionStatus::InTransactionImplicit(mut txn)
+            | TransactionStatus::InTransaction(mut txn) => {
+                if access.is_some() {
+                    txn.access = access;
```

Comment:
What about going from readwrite to readonly? That should also fail because we don't support most reads after write. Or does that effectively get handled by the set ops function? Hmm this is getting hard to reason about.

Reply:
`READ WRITE` to `READ ONLY` is permitted all the time in PG, and setting the ops handles this in cases where it should error in MZ.

Reply:
Am I understanding this correctly. The only thing that is not allowed is going from a Read Only transaction where there have already been some reads to a ReadWrite transaction? That scenario is prevented in the above check. All other scenarios are allowed which is why we don't check the old access here?

Reply:
You are only allowed to go from READ ONLY to READ WRITE if you have not yet issued a query. All other transitions are permitted (which amounts to going from READ WRITE to READ ONLY or not changing the access, both of which are always permitted). I believe the answer to your questions are "yes" and "yes."